lib/e3db/client.rb

```diff
@@ -184,30 +253,48 @@ def read(record_id)
       decrypt_record(read_raw(record_id))
     end
 
-    # Create a new, empty record that can be written to E3DB
-    # by calling {Client#write}.
+    # Write a new record to the E3DB storage service.
     #
-    # @param type [String] free-form content type of this record
-    # @return [Record] an empty record of `type`
-    def new_record(type)
+    # @param type [String] free-form content type name of this record
+    # @param data [Hash<String, String>] record data to be stored encrypted
+    # @param plain [Hash<String, String>] record data to be stored unencrypted for querying
+    # @return [Record] the newly created record object
+    def write(type, data, plain=Hash.new)
+      url = get_url('v1', 'storage', 'records')
       id = @config.client_id
       meta = Meta.new(record_id: nil, writer_id: id, user_id: id,
-                      type: type, plain: Hash.new, created: nil,
-                      last_modified: nil)
-      Record.new(meta: meta, data: Hash.new)
+                      type: type, plain: plain, created: nil,
+                      last_modified: nil, version: nil)
+      record = Record.new(meta: meta, data: data)
+      resp = @conn.post(url, encrypt_record(record).to_hash)
+      decrypt_record(Record.new(JSON.parse(resp.body, symbolize_names: true)))
     end
 
-    # Write a new record to the E3DB storage service.
+    # Update an existing record in the E3DB storage service.
     #
-    # Create new records with {Client#new_record}.
+    # If the record has been modified by another client since it was
+    # read, this method raises {ConflictError}, which should be caught
+    # by the caller so that the record can be re-fetched and the update retried.
     #
-    # @param record [Record] record to write
-    # @return [String] the unique ID of the written record
-    def write(record)
-      url = get_url('v1', 'storage', 'records')
-      resp = @conn.post(url, encrypt_record(record).to_hash)
+    # The metadata of the input record will be updated in-place to reflect
+    # the new version number and modification time returned by the server.
+    #
+    # @param record [Record] the record to update
+    def update(record)
+      record_id = record.meta.record_id
+      version = record.meta.version
+      url = get_url('v1', 'storage', 'records', 'safe', record_id, version)
+      begin
+        resp = @conn.put(url, encrypt_record(record).to_hash)
+      rescue Faraday::ClientError => e
+        if e.response[:status] == 409
+          raise E3DB::ConflictError, record
+        else
+          raise e   # re-raise on other failures
+        end
+      end
       json = JSON.parse(resp.body, symbolize_names: true)
-      json[:meta][:record_id]
+      record.meta = Meta.new(json[:meta])
```

Comment:
Ruby methods return the last statement or expression, so this should probably end with `record`.